### ISSUE TITLE
Datanodes fail to register with namenode after restarting namenode and datanode pods simultaneously 

### ISSUE THREAD

**Issue by kimoonkim** [apache-spark-on-k8s/kubernetes-HDFS#42]

**Comment by fanbowen13013** [comment]
I think it may be caused by data persistence. I using glusterfs combined statefulset to persist the data of namenode and datanode, and never met the same issue like this. Every time start the namenode will try to find the current/VERSION 
 file. If this file is missing ,namenode will not be formated. I think following link will provide some help, https://github.com/kubernetes/charts/blob/master/stable/hadoop/templates/hadoop-configmap.yaml

**Comment by fanbowen13013** [comment]
I did the same operation before, and never met this problem. Maybe following link will do some help.
https://github.com/kubernetes/charts/blob/master/stable/hadoop/templates/hadoop-configmap.yaml

**Comment by kimoonkim** [comment]
I encountered this symptom while working on the uber-chart PR #47. With the uber-chart, all subcharts including namendoes and datanodes would start simultaneoulsy. And datanodes would often fail to register with namenodes. In #47, I fixed the bug by adding a liveliness probe to datanodes. The probe would check the datanode jmx and assert it registers with namnodes. Otherwise, the probe would fail. The datanode pod would restart and eventually succeed. With the fix, I think I can close this issue.

But it seems a similar issue can happen post-registration. If namenodes resart, the datanodes would be left with stale DNS entries. I am going to file another issue for this edge case.

**Comment by xianliangz** [comment]
Hi @kimoonkim , what's the current status? We also hit similar issues when running HDFS on k8s. I feel we need to update HDFS code to make data node automatically retry the resolution of name node DNS name. Even at deployment time, if the kube-dns of name node is not ready, data node will fail to resolve the DNS, "cached" the unresolved DNS and keep retrying with the unresolved DNS. Restarting data node seems like a primitive work around.

**Comment by kimoonkim** [comment]
@xianliangz Yes, that is a known bug. Please see #48. Two approaches were discussed in the past. I may try out those approaches in the near future.
